File: memory-engine/client/memory-client.ts
/**
 * Memory Engine TypeScript Client Library
 * 
 * This library provides a strongly-typed client for interacting with the Memory Engine Service.
 */

// Type definitions
export type MemoryType = 'short_term' | 'long_term';

export interface MemoryEngineClientOptions {
  baseUrl?: string;
  apiKey?: string;
  timeout?: number;
}

export interface Memory {
  id?: string;
  userId: string;
  type: MemoryType;
  key: string;
  value: any;
  metadata?: Record<string, any>;
  ttl?: number;
  created_at?: string;
  last_accessed?: string;
}

export interface CreateMemoryResponse {
  status: string;
  id: string;
  userId: string;
  type: MemoryType;
  key: string;
}

export interface GetMemoriesResponse {
  userId: string;
  memories: Memory[];
}

export interface KnowledgeSearchParams {
  query: string;
  organizationId: string;
  limit?: number;
  threshold?: number;
}

export interface KnowledgeSearchResult {
  id: string;
  content: string;
  title?: string;
  url?: string;
  source?: string;
  relevance: number;
  metadata: Record<string, any>;
}

export interface KnowledgeSearchResponse {
  query: string;
  results: KnowledgeSearchResult[];
}

export interface VectorEmbeddingParams {
  text: string;
  metadata: Record<string, any>;
}

export interface VectorEmbeddingResponse {
  id: string;
  status: string;
}

export interface SessionContextResponse {
  sessionId: string;
  context: Record<string, any>;
}

export interface UpdateSessionContextResponse {
  sessionId: string;
  status: string;
}

export interface OptimizeContextParams {
  conversationHistory: Array<{role: string, content: string}>;
  userQuery: string;
  userId?: string;
  sessionId?: string;
  maxTokens?: number;
}

export interface OptimizedContext {
  conversation_history: Array<{role: string, content: string}>;
  relevant_memories?: Memory[];
  session_context?: Record<string, any>;
  knowledge_results?: KnowledgeSearchResult[];
}

export interface OptimizeContextResponse {
  optimized_context: OptimizedContext;
  token_count: number;
}

export interface HealthCheckResponse {
  status: string;
  version?: string;
  dependencies: {
    redis: string;
    chroma: string;
    openai: string;
  };
}

/**
 * Memory Engine Client for TypeScript applications
 */
export class MemoryEngineClient {
  private baseUrl: string;
  private apiKey?: string;
  private timeout: number;
  private headers: Record<string, string>;

  /**
   * Initialize the Memory Engine Client
   * 
   * @param options - Client options
   */
  constructor(options: MemoryEngineClientOptions = {}) {
    this.baseUrl = options.baseUrl || 'http://localhost:8000';
    this.apiKey = options.apiKey;
    this.timeout = options.timeout || 10000;
    this.headers = {
      'Content-Type': 'application/json',
    };

    if (this.apiKey) {
      this.headers['X-API-Key'] = this.apiKey;
    }
  }

  /**
   * Create a new memory or update an existing one
   * 
   * @param memory - Memory object
   * @returns Response from the Memory Engine Service
   */
  async createMemory(memory: Memory): Promise<CreateMemoryResponse> {
    const url = `${this.baseUrl}/memories`;
    
    try {
      const response = await this.fetchWithTimeout(url, {
        method: 'POST',
        headers: this.headers,
        body: JSON.stringify(memory),
      });

      return await response.json();
    } catch (error) {
      console.error('Error creating memory:', error);
      throw error;
    }
  }

  /**
   * Retrieve memories for a user
   * 
   * @param userId - User ID
   * @param type - Optional memory type filter
   * @param key - Optional key filter
   * @returns Response from the Memory Engine Service
   */
  async getMemories(userId: string, type?: MemoryType, key?: string): Promise<GetMemoriesResponse> {
    let url = `${this.baseUrl}/memories/${userId}`;
    const params = new URLSearchParams();

    if (type) {
      params.append('type', type);
    }

    if (key) {
      params.append('key', key);
    }

    if (params.toString()) {
      url += `?${params.toString()}`;
    }

    try {
      const response = await this.fetchWithTimeout(url, {
        method: 'GET',
        headers: this.headers,
      });

      return await response.json();
    } catch (error) {
      console.error('Error retrieving memories:', error);
      throw error;
    }
  }

  /**
   * Search the knowledge base for relevant information
   * 
   * @param params - Search parameters
   * @returns Response from the Memory Engine Service
   */
  async searchKnowledge(params: KnowledgeSearchParams): Promise<KnowledgeSearchResponse> {
    const url = `${this.baseUrl}/knowledge/search`;
    
    try {
      const response = await this.fetchWithTimeout(url, {
        method: 'POST',
        headers: this.headers,
        body: JSON.stringify(params),
      });

      return await response.json();
    } catch (error) {
      console.error('Error searching knowledge base:', error);
      throw error;
    }
  }

  /**
   * Store a vector embedding for future similarity search
   * 
   * @param params - Vector parameters
   * @returns Response from the Memory Engine Service
   */
  async createVectorEmbedding(params: VectorEmbeddingParams): Promise<VectorEmbeddingResponse> {
    const url = `${this.baseUrl}/vectors`;
    
    try {
      const response = await this.fetchWithTimeout(url, {
        method: 'POST',
        headers: this.headers,
        body: JSON.stringify(params),
      });

      return await response.json();
    } catch (error) {
      console.error('Error creating vector embedding:', error);
      throw error;
    }
  }

  /**
   * Get the context for a specific session
   * 
   * @param sessionId - Session ID
   * @returns Response from the Memory Engine Service
   */
  async getSessionContext(sessionId: string): Promise<SessionContextResponse> {
    const url = `${this.baseUrl}/context/${sessionId}`;

    try {
      const response = await this.fetchWithTimeout(url, {
        method: 'GET',
        headers: this.headers,
      });

      return await response.json();
    } catch (error) {
      console.error('Error retrieving session context:', error);
      throw error;
    }
  }

  /**
   * Update the context for a specific session
   * 
   * @param sessionId - Session ID
   * @param context - Session context
   * @returns Response from the Memory Engine Service
   */
  async updateSessionContext(sessionId: string, context: Record<string, any>): Promise<UpdateSessionContextResponse> {
    const url = `${this.baseUrl}/context/${sessionId}`;

    try {
      const response = await this.fetchWithTimeout(url, {
        method: 'POST',
        headers: this.headers,
        body: JSON.stringify(context),
      });

      return await response.json();
    } catch (error) {
      console.error('Error updating session context:', error);
      throw error;
    }
  }

  /**
   * Optimize the context window for token efficiency
   * 
   * @param params - Context parameters
   * @returns Response from the Memory Engine Service
   */
  async optimizeContext(params: OptimizeContextParams): Promise<OptimizeContextResponse> {
    const url = `${this.baseUrl}/optimize-context`;
    const payload = {
      conversation_history: params.conversationHistory,
      user_query: params.userQuery,
      max_tokens: params.maxTokens || 4000,
    } as Record<string, any>;

    if (params.userId) {
      payload.userId = params.userId;
    }

    if (params.sessionId) {
      payload.sessionId = params.sessionId;
    }

    try {
      const response = await this.fetchWithTimeout(url, {
        method: 'POST',
        headers: this.headers,
        body: JSON.stringify(payload),
      });

      return await response.json();
    } catch (error) {
      console.error('Error optimizing context:', error);
      throw error;
    }
  }

  /**
   * Check the health status of the Memory Engine Service
   * 
   * @returns Response from the Memory Engine Service
   */
  async healthCheck(): Promise<HealthCheckResponse> {
    const url = `${this.baseUrl}/health`;

    try {
      const response = await this.fetchWithTimeout(url, {
        method: 'GET',
      });

      return await response.json();
    } catch (error) {
      console.error('Health check failed:', error);
      throw error;
    }
  }

  /**
   * Fetch with timeout
   * 
   * @private
   * @param url - URL to fetch
   * @param options - Fetch options
   * @returns Fetch response
   */
  private async fetchWithTimeout(url: string, options: RequestInit = {}): Promise<Response> {
    const controller = new AbortController();
    const id = setTimeout(() => controller.abort(), this.timeout);
    
    try {
      const response = await fetch(url, {
        ...options,
        signal: controller.signal,
      });
      
      clearTimeout(id);
      
      if (!response.ok) {
        const errorText = await response.text();
        throw new Error(`HTTP error ${response.status}: ${errorText}`);
      }
      
      return response;
    } catch (error) {
      clearTimeout(id);
      
      if (error instanceof Error && error.name === 'AbortError') {
        throw new Error(`Request timeout after ${this.timeout}ms`);
      }
      
      throw error;
    }
  }
}

// Export for CommonJS and ES modules
export default MemoryEngineClient;

File: memory-engine/client/memory_client.py
#!/usr/bin/env python3
"""
Memory Engine Client Library

This library provides a simple client for interacting with the Memory Engine Service.
"""

import json
import logging
from typing import Dict, List, Optional, Any, Union

import httpx

logger = logging.getLogger("memory_client")


class MemoryEngineClient:
    """Client for interacting with the Memory Engine Service."""

    def __init__(
        self,
        base_url: str = "http://localhost:8000",
        api_key: str = None,
        timeout: int = 10
    ):
        """
        Initialize the Memory Engine Client.

        Args:
            base_url: Base URL of the Memory Engine Service
            api_key: API key for authentication
            timeout: Request timeout in seconds
        """
        self.base_url = base_url
        self.api_key = api_key
        self.timeout = timeout
        self.headers = {"Content-Type": "application/json"}
        
        if api_key:
            self.headers["X-API-Key"] = api_key

    async def create_memory(
        self,
        user_id: str,
        memory_type: str,
        key: str,
        value: Dict[str, Any],
        metadata: Optional[Dict[str, Any]] = None,
        ttl: Optional[int] = None
    ) -> Dict[str, Any]:
        """
        Create a new memory or update an existing one.

        Args:
            user_id: User ID
            memory_type: Memory type ('short_term' or 'long_term')
            key: Unique identifier for this memory type
            value: Memory content
            metadata: Additional metadata
            ttl: Time to live in seconds (for short_term only)

        Returns:
            Response from the Memory Engine Service
        """
        url = f"{self.base_url}/memories"
        payload = {
            "userId": user_id,
            "type": memory_type,
            "key": key,
            "value": value,
            "metadata": metadata or {},
        }
        
        if ttl is not None:
            payload["ttl"] = ttl

        async with httpx.AsyncClient(timeout=self.timeout) as client:
            response = await client.post(
                url,
                headers=self.headers,
                json=payload
            )
            
            if response.status_code != 200:
                logger.error(f"Error creating memory: {response.text}")
                response.raise_for_status()
                
            return response.json()

    async def get_memories(
        self,
        user_id: str,
        memory_type: Optional[str] = None,
        key: Optional[str] = None
    ) -> Dict[str, Any]:
        """
        Retrieve memories for a user.

        Args:
            user_id: User ID
            memory_type: Optional memory type filter
            key: Optional key filter

        Returns:
            Response from the Memory Engine Service
        """
        url = f"{self.base_url}/memories/{user_id}"
        params = {}
        
        if memory_type:
            params["type"] = memory_type
            
        if key:
            params["key"] = key

        async with httpx.AsyncClient(timeout=self.timeout) as client:
            response = await client.get(
                url,
                headers=self.headers,
                params=params
            )
            
            if response.status_code != 200:
                logger.error(f"Error retrieving memories: {response.text}")
                response.raise_for_status()
                
            return response.json()

    async def search_knowledge(
        self,
        query: str,
        organization_id: str,
        limit: int = 5,
        threshold: float = 0.7
    ) -> Dict[str, Any]:
        """
        Search the knowledge base for relevant information.

        Args:
            query: Search query
            organization_id: Organization ID
            limit: Maximum number of results
            threshold: Minimum relevance threshold

        Returns:
            Response from the Memory Engine Service
        """
        url = f"{self.base_url}/knowledge/search"
        payload = {
            "query": query,
            "organizationId": organization_id,
            "limit": limit,
            "threshold": threshold
        }

        async with httpx.AsyncClient(timeout=self.timeout) as client:
            response = await client.post(
                url,
                headers=self.headers,
                json=payload
            )
            
            if response.status_code != 200:
                logger.error(f"Error searching knowledge base: {response.text}")
                response.raise_for_status()
                
            return response.json()

    async def create_vector_embedding(
        self,
        text: str,
        metadata: Dict[str, Any]
    ) -> Dict[str, Any]:
        """
        Store a vector embedding for future similarity search.

        Args:
            text: Text to embed
            metadata: Metadata for the embedding

        Returns:
            Response from the Memory Engine Service
        """
        url = f"{self.base_url}/vectors"
        payload = {
            "text": text,
            "metadata": metadata
        }

        async with httpx.AsyncClient(timeout=self.timeout) as client:
            response = await client.post(
                url,
                headers=self.headers,
                json=payload
            )
            
            if response.status_code != 200:
                logger.error(f"Error creating vector embedding: {response.text}")
                response.raise_for_status()
                
            return response.json()

    async def get_session_context(
        self,
        session_id: str
    ) -> Dict[str, Any]:
        """
        Get the context for a specific session.

        Args:
            session_id: Session ID

        Returns:
            Response from the Memory Engine Service
        """
        url = f"{self.base_url}/context/{session_id}"

        async with httpx.AsyncClient(timeout=self.timeout) as client:
            response = await client.get(
                url,
                headers=self.headers
            )
            
            if response.status_code != 200:
                logger.error(f"Error retrieving session context: {response.text}")
                response.raise_for_status()
                
            return response.json()

    async def update_session_context(
        self,
        session_id: str,
        context: Dict[str, Any]
    ) -> Dict[str, Any]:
        """
        Update the context for a specific session.

        Args:
            session_id: Session ID
            context: Session context

        Returns:
            Response from the Memory Engine Service
        """
        url = f"{self.base_url}/context/{session_id}"

        async with httpx.AsyncClient(timeout=self.timeout) as client:
            response = await client.post(
                url,
                headers=self.headers,
                json=context
            )
            
            if response.status_code != 200:
                logger.error(f"Error updating session context: {response.text}")
                response.raise_for_status()
                
            return response.json()

    async def optimize_context(
        self,
        conversation_history: List[Dict[str, str]],
        user_query: str,
        user_id: Optional[str] = None,
        session_id: Optional[str] = None,
        max_tokens: int = 4000
    ) -> Dict[str, Any]:
        """
        Optimize the context window for token efficiency.

        Args:
            conversation_history: Conversation history
            user_query: User query
            user_id: Optional user ID
            session_id: Optional session ID
            max_tokens: Maximum tokens

        Returns:
            Response from the Memory Engine Service
        """
        url = f"{self.base_url}/optimize-context"
        payload = {
            "conversation_history": conversation_history,
            "user_query": user_query,
            "max_tokens": max_tokens
        }
        
        if user_id:
            payload["userId"] = user_id
            
        if session_id:
            payload["sessionId"] = session_id

        async with httpx.AsyncClient(timeout=self.timeout) as client:
            response = await client.post(
                url,
                headers=self.headers,
                json=payload
            )
            
            if response.status_code != 200:
                logger.error(f"Error optimizing context: {response.text}")
                response.raise_for_status()
                
            return response.json()

    async def health_check(self) -> Dict[str, Any]:
        """
        Check the health status of the Memory Engine Service.

        Returns:
            Response from the Memory Engine Service
        """
        url = f"{self.base_url}/health"

        async with httpx.AsyncClient(timeout=self.timeout) as client:
            response = await client.get(url)
            
            if response.status_code != 200:
                logger.error(f"Health check failed: {response.text}")
                response.raise_for_status()
                
            return response.json()


# Synchronous client for non-async environments
class SyncMemoryEngineClient:
    """Synchronous client for interacting with the Memory Engine Service."""

    def __init__(
        self,
        base_url: str = "http://localhost:8000",
        api_key: str = None,
        timeout: int = 10
    ):
        """
        Initialize the Synchronous Memory Engine Client.

        Args:
            base_url: Base URL of the Memory Engine Service
            api_key: API key for authentication
            timeout: Request timeout in seconds
        """
        self.base_url = base_url
        self.api_key = api_key
        self.timeout = timeout
        self.headers = {"Content-Type": "application/json"}
        
        if api_key:
            self.headers["X-API-Key"] = api_key

    def create_memory(
        self,
        user_id: str,
        memory_type: str,
        key: str,
        value: Dict[str, Any],
        metadata: Optional[Dict[str, Any]] = None,
        ttl: Optional[int] = None
    ) -> Dict[str, Any]:
        """
        Create a new memory or update an existing one.

        Args:
            user_id: User ID
            memory_type: Memory type ('short_term' or 'long_term')
            key: Unique identifier for this memory type
            value: Memory content
            metadata: Additional metadata
            ttl: Time to live in seconds (for short_term only)

        Returns:
            Response from the Memory Engine Service
        """
        url = f"{self.base_url}/memories"
        payload = {
            "userId": user_id,
            "type": memory_type,
            "key": key,
            "value": value,
            "metadata": metadata or {},
        }
        
        if ttl is not None:
            payload["ttl"] = ttl

        with httpx.Client(timeout=self.timeout) as client:
            response = client.post(
                url,
                headers=self.headers,
                json=payload
            )
            
            if response.status_code != 200:
                logger.error(f"Error creating memory: {response.text}")
                response.raise_for_status()
                
            return response.json()

    def get_memories(
        self,
        user_id: str,
        memory_type: Optional[str] = None,
        key: Optional[str] = None
    ) -> Dict[str, Any]:
        """
        Retrieve memories for a user.

        Args:
            user_id: User ID
            memory_type: Optional memory type filter
            key: Optional key filter

        Returns:
            Response from the Memory Engine Service
        """
        url = f"{self.base_url}/memories/{user_id}"
        params = {}
        
        if memory_type:
            params["type"] = memory_type
            
        if key:
            params["key"] = key

        with httpx.Client(timeout=self.timeout) as client:
            response = client.get(
                url,
                headers=self.headers,
                params=params
            )
            
            if response.status_code != 200:
                logger.error(f"Error retrieving memories: {response.text}")
                response.raise_for_status()
                
            return response.json()

    def search_knowledge(
        self,
        query: str,
        organization_id: str,
        limit: int = 5,
        threshold: float = 0.7
    ) -> Dict[str, Any]:
        """
        Search the knowledge base for relevant information.

        Args:
            query: Search query
            organization_id: Organization ID
            limit: Maximum number of results
            threshold: Minimum relevance threshold

        Returns:
            Response from the Memory Engine Service
        """
        url = f"{self.base_url}/knowledge/search"
        payload = {
            "query": query,
            "organizationId": organization_id,
            "limit": limit,
            "threshold": threshold
        }

        with httpx.Client(timeout=self.timeout) as client:
            response = client.post(
                url,
                headers=self.headers,
                json=payload
            )
            
            if response.status_code != 200:
                logger.error(f"Error searching knowledge base: {response.text}")
                response.raise_for_status()
                
            return response.json()

    def create_vector_embedding(
        self,
        text: str,
        metadata: Dict[str, Any]
    ) -> Dict[str, Any]:
        """
        Store a vector embedding for future similarity search.

        Args:
            text: Text to embed
            metadata: Metadata for the embedding

        Returns:
            Response from the Memory Engine Service
        """
        url = f"{self.base_url}/vectors"
        payload = {
            "text": text,
            "metadata": metadata
        }

        with httpx.Client(timeout=self.timeout) as client:
            response = client.post(
                url,
                headers=self.headers,
                json=payload
            )
            
            if response.status_code != 200:
                logger.error(f"Error creating vector embedding: {response.text}")
                response.raise_for_status()
                
            return response.json()

    def get_session_context(
        self,
        session_id: str
    ) -> Dict[str, Any]:
        """
        Get the context for a specific session.

        Args:
            session_id: Session ID

        Returns:
            Response from the Memory Engine Service
        """
        url = f"{self.base_url}/context/{session_id}"

        with httpx.Client(timeout=self.timeout) as client:
            response = client.get(
                url,
                headers=self.headers
            )
            
            if response.status_code != 200:
                logger.error(f"Error retrieving session context: {response.text}")
                response.raise_for_status()
                
            return response.json()

    def update_session_context(
        self,
        session_id: str,
        context: Dict[str, Any]
    ) -> Dict[str, Any]:
        """
        Update the context for a specific session.

        Args:
            session_id: Session ID
            context: Session context

        Returns:
            Response from the Memory Engine Service
        """
        url = f"{self.base_url}/context/{session_id}"

        with httpx.Client(timeout=self.timeout) as client:
            response = client.post(
                url,
                headers=self.headers,
                json=context
            )
            
            if response.status_code != 200:
                logger.error(f"Error updating session context: {response.text}")
                response.raise_for_status()
                
            return response.json()

    def optimize_context(
        self,
        conversation_history: List[Dict[str, str]],
        user_query: str,
        user_id: Optional[str] = None,
        session_id: Optional[str] = None,
        max_tokens: int = 4000
    ) -> Dict[str, Any]:
        """
        Optimize the context window for token efficiency.

        Args:
            conversation_history: Conversation history
            user_query: User query
            user_id: Optional user ID
            session_id: Optional session ID
            max_tokens: Maximum tokens

        Returns:
            Response from the Memory Engine Service
        """
        url = f"{self.base_url}/optimize-context"
        payload = {
            "conversation_history": conversation_history,
            "user_query": user_query,
            "max_tokens": max_tokens
        }
        
        if user_id:
            payload["userId"] = user_id
            
        if session_id:
            payload["sessionId"] = session_id

        with httpx.Client(timeout=self.timeout) as client:
            response = client.post(
                url,
                headers=self.headers,
                json=payload
            )
            
            if response.status_code != 200:
                logger.error(f"Error optimizing context: {response.text}")
                response.raise_for_status()
                
            return response.json()

    def health_check(self) -> Dict[str, Any]:
        """
        Check the health status of the Memory Engine Service.

        Returns:
            Response from the Memory Engine Service
        """
        url = f"{self.base_url}/health"

        with httpx.Client(timeout=self.timeout) as client:
            response = client.get(url)
            
            if response.status_code != 200:
                logger.error(f"Health check failed: {response.text}")
                response.raise_for_status()
                
            return response.json()
